timemachine/src/time_travel.rs

```diff
@@ -0,0 +1,742 @@
+          rusqlite::params![&key.format_version, &key.monoseq, &key.seq],
+        )?;
+        tx.commit()?;
+      }
+    }
+
+    // Step 3: Fetch REDO logs until DV is reached and we are up to date as measured
+    // in real time
+    let start_ts = SystemTime::now()
+      .duration_since(SystemTime::UNIX_EPOCH)
+      .unwrap()
+      .as_millis() as u64;
+    let last_key = self
+      .fetch_redo_outside_tx(source, |x| {
+        x.key.first_versionstamp12 > dv && x.last_modified_ms > start_ts
+      })
+      .await?;
+
+    // Step 4: Apply REDO to bring kv_snapshot to DV
+    {
+      let tx = self.db.transaction()?;
+      if get_config(&tx, TT_CONFIG_KEY_INITIAL_SNAPSHOT_COMPLETED)?.as_deref()
+        == Some("1")
+      {
+        return Ok(());
+      }
+
+      set_config(
+        &tx,
+        TT_CONFIG_KEY_CURRENT_SNAPSHOT_VERSIONSTAMP12,
+        &hex::encode([0u8; 12]),
+      )?;
+      if last_key.is_some() {
+        replay_redo_or_undo_in_tx(&tx, dv, false)?;
+      }
+      set_config(&tx, TT_CONFIG_KEY_INITIAL_SNAPSHOT_COMPLETED, "1")?;
+      tx.commit()?;
+    }
+
+    Ok(())
+  }
+
+  pub async fn sync(
+    &mut self,
+    source: &impl DatabaseBackupSource,
+  ) -> Result<()> {
+    if !self.initial_snapshot_is_complete()? {
+      anyhow::bail!("initial snapshot not completed");
+    }
+    let start_ts = SystemTime::now()
+      .duration_since(SystemTime::UNIX_EPOCH)
+      .unwrap()
+      .as_millis() as u64;
+    self
+      .fetch_redo_outside_tx(source, |x| x.last_modified_ms > start_ts)
+      .await?;
+    Ok(())
+  }
+
+  pub fn get_redo_cursor(&self) -> Result<Option<MutationRangeKey>> {
+    query_redo_cursor(&self.db)
+  }
+
+  pub fn get_s3_config(
+    &mut self,
+  ) -> Result<Option<DatabaseBackupSourceS3Config>> {
+    let tx = self.db.transaction()?;
+    let s3_bucket = get_config(&tx, TT_CONFIG_KEY_S3_BUCKET)?;
+    let s3_prefix = get_config(&tx, TT_CONFIG_KEY_S3_PREFIX)?;
+
+    let (Some(s3_bucket), Some(s3_prefix)) = (s3_bucket, s3_prefix) else {
+      return Ok(None);
+    };
+
+    Ok(Some(DatabaseBackupSourceS3Config {
+      bucket: s3_bucket,
+      prefix: s3_prefix,
+    }))
+  }
+
+  pub fn init_s3(
+    &mut self,
+    s3_config: &DatabaseBackupSourceS3Config,
+  ) -> Result<()> {
+    let tx = self.db.transaction()?;
+    if get_config(&tx, TT_CONFIG_KEY_S3_INIT_COMPLETED)?.as_deref() == Some("1")
+    {
+      // Validate consistency
+      let current_bucket =
+        get_config(&tx, TT_CONFIG_KEY_S3_BUCKET)?.unwrap_or_default();
+      let current_prefix =
+        get_config(&tx, TT_CONFIG_KEY_S3_PREFIX)?.unwrap_or_default();
+
+      if current_bucket != s3_config.bucket
+        || current_prefix != s3_config.prefix
+      {
+        anyhow::bail!(
+          "Database is previously synced with a different S3 bucket or prefix. If you have moved the backup directory, please use the SQLite CLI to manually delete `s3_init_completed` in table `tt_config`."
+        );
```

Comment:
Can we just add a `--force` flag for this?

Reply:
This is very dangerous and should not made too easy :)